### PR TITLE
Complete fix for issue 656

### PR DESCRIPTION
**Pull request by bgavrilMS**
https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/656

### CODE REVIEW

File: core/src/Platforms/Android/AndroidPlatformProxy.cs [hunk 1]
```diff
@@ -135,11 +135,21 @@ public string GetProductName()
             return _isMsal ? "MSAL.Xamarin.Android" : "PCL.Android";
         }
 
+        private readonly ILegacyCachePersistence createLegacyCachePersistence = new AndroidLegacyCachePersistence();
```
Comment (MarkZuber) [inline core/src/Platforms/Android/AndroidPlatformProxy.cs]: Why on android is the value re-used when it's created new on every other platform?

File: msal/tests/Test.MSAL.NET.Unit.net45/CoreTests/PlatformProxyFactoryTests.cs [hunk 2]
```diff
@@ -43,5 +43,26 @@ public void PlatformProxyFactoryCachesTheProxy()
...
+            // Act and Assert
+            Assert.AreNotEqual(
+                factory.CreateLegacyCachePersistence(),
+                factory.CreateLegacyCachePersistence());
+
+            Assert.AreNotEqual(
+                factory.CreateTokenCacheAccessor(),
+                factory.CreateTokenCacheAccessor());
+
+            Assert.AreEqual(
```
Comment (MarkZuber) [inline msal/tests/Test.MSAL.NET.Unit.net45/CoreTests/PlatformProxyFactoryTests.cs]: Do you want "AreEqual" here or "AreSame" (which checks for the same reference)

**Review by MarkZuber — CHANGES_REQUESTED** [review]
🕐

File: msal/tests/Test.MSAL.NET.Unit.net45/CoreTests/PlatformProxyFactoryTests.cs (hunk 2 above)
Comment (bgavrilMS) [reply msal/tests/Test.MSAL.NET.Unit.net45/CoreTests/PlatformProxyFactoryTests.cs]: fixed.

File: core/src/Platforms/Android/AndroidPlatformProxy.cs (hunk 1 above)
Comment (bgavrilMS) [reply core/src/Platforms/Android/AndroidPlatformProxy.cs]: Good catch, fixed.

**Review by MarkZuber — APPROVED** [review]

File: core/src/Platforms/Android/AndroidPlatformProxy.cs (hunk 1 above)
Comment (henrik-me) [reply core/src/Platforms/Android/AndroidPlatformProxy.cs]: as far as I can see this is still not fixed (uniform across languages)

and also why would it need to be created new for every call? Name is meaningful, however isn't there a need to retrieve the one created?

---
In reply to: [229044749](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1364#discussion_r229044749) [](ancestors = 229044749)

**Comment by henrik-me** [comment]
See also: [https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1360](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1360)

**Review by henrik-me — CHANGES_REQUESTED** [review]
:clock1:

File: core/src/Platforms/Android/AndroidPlatformProxy.cs (hunk 1 above)
Comment (bgavrilMS) [reply core/src/Platforms/Android/AndroidPlatformProxy.cs]: LegacyCachePersistance is not a singleton, a separate instance is created with each TokenCache. Also on some platforms, such as net45, it has state, so reusing the same instance will have consequences.

**Pull request merged**